SCC hostNetwork/hostPort support

pweil- opened the pull request:
upstream: https://github.com/GoogleCloudPlatform/kubernetes/pull/7893

Allow controlling `hostNetwork` and `hostPort` with an SCC.  

Note: this still works with `HostNetworkSources` defined in `capabilities.go`.  If the pod source is not allowed by `HostNetworkSources` then the kubelet check in `canRunPod` will fail even if the SCC allows the request.  I think it is probably best to leave the kubelet check in place for now.  In the future I think it's reasonable to remove `HostNetworkSources` and allow the kubelet to assume that admission has performed the necessary checks and restricting sources becomes allocating permissions to users/groups in the SCC.


@liggitt @pmorie @smarterclayton

smarterclayton commented:
On Jun 29, 2015, at 2:25 PM, Paul Weil <notifications@github.com> wrote:

upstream: GoogleCloudPlatform/kubernetes#7893
<https://github.com/GoogleCloudPlatform/kubernetes/pull/7893>

Allow controlling pod.spec.hostNetwork with an SCC.

Note: this still works with HostNetworkSources defined in capabilities.go.
If the pod source is not allowed by HostNetworkSources then the kubelet
check in canRunPod will fail even if the SCC allows the request. I think it
is probably best to leave the kubelet check in place for now. In the future
I think it's reasonable to remove HostNetworkSources and allow the kubelet
to assume that admission has performed the necessary checks and restricting
sources becomes allocating permissions to users/groups in the SCC.

I think we may have a future use case that desires a node to be able to
reject external pods that don't match an SCC - so if we remove host network
sources we need an SCC on the kubelet (maybe)

# Ensure you may not use host network with the default set of SCCs
[vagrant@openshiftdev scc]$ oc create -f hello-pod-hostnetwork.json
Error from server: Pod "hello-openshift" is forbidden: unable to
validate against any security context constraint: [provider
restricted: .spec.containers[0].securityContext.hostNetwork: invalid
value 'true': Host network is not allowed to be used provider
privileged: .spec.containers[0].securityContext.hostNetwork: invalid
value 'true': Host network is not allowed to be used]

# Edit the privileged SCC to allow host network
[vagrant@openshiftdev scc]$ oc edit scc privileged
securitycontextconstraints/privileged

# Successful create
[vagrant@openshiftdev scc]$ oc create -f hello-pod-hostnetwork.json
pods/hello-openshift

# Check settings - validates against priv scc and has hostNetwork = true
[vagrant@openshiftdev scc]$ oc get pod hello-openshift -o json | grep
"scc\|hostNetwork"
            "openshift.io/scc": "privileged"
        "hostNetwork": true,

# Test running pod
[vagrant@openshiftdev scc]$ curl localhost:8080
Hello OpenShift!

@liggitt <https://github.com/liggitt> @pmorie <https://github.com/pmorie>
@smarterclayton <https://github.com/smarterclayton>
------------------------------
You can view, comment on, or merge this pull request online at:

  https://github.com/openshift/origin/pull/3496
Commit Summary

   - UPSTREAM: SCC hostNetwork
   https://github.com/GoogleCloudPlatform/kubernetes/pull/7893

File Changes

   - *M*
   Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/types.go
   <https://github.com/openshift/origin/pull/3496/files#diff-0> (2)
   - *M*
   Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1/conversion_generated.go
   <https://github.com/openshift/origin/pull/3496/files#diff-1> (2)
   - *M*
   Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1/types.go
   <https://github.com/openshift/origin/pull/3496/files#diff-2> (2)
   - *M*
   Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1beta1/types.go
   <https://github.com/openshift/origin/pull/3496/files#diff-3> (2)
   - *M*
   Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1beta2/types.go
   <https://github.com/openshift/origin/pull/3496/files#diff-4> (2)
   - *M*
   Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1beta3/conversion_generated.go
   <https://github.com/openshift/origin/pull/3496/files#diff-5> (2)
   - *M*
   Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1beta3/types.go
   <https://github.com/openshift/origin/pull/3496/files#diff-6> (2)
   - *M*
   Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/securitycontextconstraints/provider.go
   <https://github.com/openshift/origin/pull/3496/files#diff-7> (4)
   - *M*
   Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/securitycontextconstraints/provider_test.go
   <https://github.com/openshift/origin/pull/3496/files#diff-8> (19)
   - *M* pkg/security/admission/admission_test.go
   <https://github.com/openshift/origin/pull/3496/files#diff-9> (8)

Patch Links:

   - https://github.com/openshift/origin/pull/3496.patch
   - https://github.com/openshift/origin/pull/3496.diff

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3496>.

pweil- commented:
updating to [DO NOT MERGE] until rebase is completed, this will then apply as a clean patch to the rebased k8s master.

pweil- commented:
rebased to master

smarterclayton commented:
@derekwaynecarr or @pmorie review for merge.

On Jul 9, 2015, at 9:29 AM, Paul Weil <notifications@github.com> wrote:

rebased to master

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3496#issuecomment-119965562>.

derekwaynecarr commented on Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/types.go:
```diff
@@ -2166,11 +2166,15 @@ type SecurityContextConstraints struct {
 
 	// AllowPrivilegedContainer determines if a container can request to be run as privileged.
-	AllowPrivilegedContainer bool `json:"allowPrivilegedContainer,omitempty" description:"allow containers to run as privileged"`
+	AllowPrivilegedContainer bool `json:"allowPrivilegedContainer" description:"allow containers to run as privileged"`
```
we should not have `json:...` annotations in pkg/api/types.go, only in versioned apis.

pweil- replied:
10-4, will remove these

smarterclayton commented:
You have origin changes mixed in with your Kube commit

smarterclayton commented:
Fix that and this is LGTM and we can merge

smarterclayton commented:
Is there an open issue to flip the router to host network?

pweil- commented:
@derekwaynecarr - removed json notation from internal api types
@smarterclayton - broken into separate commits

pweil- commented:
> Is there an open issue to flip the router to host network?

Not an open issue but we have some of the pieces already R&D'd.  I'll open one

Ref issues: https://github.com/openshift/origin/pull/3299, https://github.com/openshift/origin/pull/3297

smarterclayton commented:
LGTM - derek, any other comments?

pweil- commented:
@derekwaynecarr - any other feedback here?

smarterclayton commented:
Derek's out.  [merge].  You have a card/issue for moving SCC to OpenShift for now?

pweil- commented:
> You have a card/issue for moving SCC to OpenShift for now?

Yes, a card has been created for this.

smarterclayton commented:
Link here please

On Mon, Jul 20, 2015 at 1:43 PM, Paul Weil <notifications@github.com> wrote:

> You have a card/issue for moving SCC to OpenShift for now?
>
> Yes, a card has been created for this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3496#issuecomment-122964711>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

pweil- commented:
https://trello.com/c/rLoMBUH1/472-5-refactor-scc-model-into-openshift-types

pweil- commented:
hrm, guessing that e2e failure is the router using host ports but only having access to the restricted SCC.  Well played jenkins, well played.  Will look into this.

pweil- commented:
[test]

pweil- commented:
So, we are basically now requiring that the router be run under an escalated scc with this since it uses host ports.  To help with this, I've updated the router admin command to require a service account and it will check that the service account specified has access to a host port enabled SCC.  I've added some code to the e2e test to support this.

Strangely, the vagrant env was getting device errors running e2e that needs investigated.

@smarterclayton @pmorie - ptal and let me know if there are concerns with the above approach.  If not, when the light is green for e2e this is ready for re-merge.

smarterclayton commented:
That's ok - the router will go to hostNetwork and need the same priv.

On Jul 21, 2015, at 10:15 PM, Paul Weil <notifications@github.com> wrote:

So, we are basically now requiring that the router be run under an
escalated scc with this since it uses host ports. To help with this, I've
updated the router admin command to require a service account and it will
check that the service account specified has access to a host port enabled
SCC. I've added some code to the e2e test to support this.

Strangely, the vagrant env was getting device errors running e2e that needs
investigated.

@smarterclayton <https://github.com/smarterclayton> @pmorie
<https://github.com/pmorie> - ptal and let me know if there are concerns
with the above approach. If not, when the light is green for e2e this is
ready for re-merge.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3496#issuecomment-123535834>.

smarterclayton commented:
LGTM - merge at will

pweil- commented:
re-[test]

pweil- commented:
one more go-round for docker int tests [test]

pmorie commented on Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/types.go:
```diff
@@ -2162,45 +2162,49 @@ type RangeAllocation struct {
+	AllowedCapabilities []Capability
 	// AllowHostDirVolumePlugin determines if the policy allow containers to use the HostDir volume plugin
-	AllowHostDirVolumePlugin bool `json:"allowHostDirVolumePlugin,omitempty" description:"allow the use of the host dir volume plugin"`
+	AllowHostDirVolumePlugin bool
```
You might consider making the comment changes in another PR -- I had one that did this, which is closed, but this seems like something someone might just go and do and then could fall out as a rebase problem in the future.

smarterclayton replied:
Hurry - want to get this merged!

On Wed, Jul 22, 2015 at 1:27 PM, Paul Morie <notifications@github.com>
wrote:

> In
> Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/types.go
> <https://github.com/openshift/origin/pull/3496#discussion_r35239350>:
>
> >  	// AllowHostDirVolumePlugin determines if the policy allow containers to use the HostDir volume plugin
> > -	AllowHostDirVolumePlugin bool `json:"allowHostDirVolumePlugin,omitempty" description:"allow the use of the host dir volume plugin"`
> > +	AllowHostDirVolumePlugin bool
>
> You might consider making the comment changes in another PR -- I had one
> that did this, which is closed, but this seems like something someone might
> just go and do and then could fall out as a rebase problem in the future.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3496/files#r35239350>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

pweil- replied:
Changes to this won't come from a rebase since the SCC type isn't in upstream.  This would just build upon the cherry-picks but hopefully we'll remove the need for that soon with the refactor.

pmorie replied:
Good point, my bad.

pmorie commented on pkg/cmd/admin/router/router.go:
```diff
@@ -225,6 +230,15 @@ func RunCmdRouter(f *clientcmd.Factory, cmd *cobra.Command, out io.Writer, cfg *
 		}
 
+		if len(cfg.ServiceAccount) == 0 {
+			return fmt.Errorf("router could not be created; you must specify a service account with access to host ports with --service-account")
```
I don't think this error message should say anything about host ports -- the problem is that no service account is specified

pmorie commented:
Ok, LGTM aside from the one thing about the error message - fix that and I will tag.

pweil- commented:
@pmorie message updated to `"router could not be created; you must specify a service account with --service-account"`

pmorie commented:
@pweil- cool

pmorie commented:
[merge]

rhcarvalho commented:
@pweil- could you please update the instructions on how to create a router in the sample-app [README](https://github.com/openshift/origin/blob/master/examples/sample-app/README.md#advanced)?

pweil- commented:
@rhcarvalho - thanks for pointing this out, I would've missed it.  Will update

liggitt commented:
This has compatibility issues, since a service account wasn't required previously. Is hostNetwork required now?  This is at least worth an announcement email, and the upgrade read me needs to be updated.  I'm pretty sure it broke the quick installer also

liggitt commented:
Could the service account arg default to "default"?

smarterclayton commented:
Is it even really required?  If empty and no credentials, assume the user
is using a service account?

On Aug 3, 2015, at 7:45 PM, Jordan Liggitt <notifications@github.com> wrote:

Could the service account arg default to "default"?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3496#issuecomment-127434396>.

liggitt commented:
The command does validation to make sure the chosen service account has the required SCC permissions. I'd rather let the pod creation do that validation, myself (and if errors aren't surfaced clearly enough, work on improving that). If they want to change the chosen service account after the fact, it's as simple as editing the router replication controller.

liggitt commented:
We should also prioritize making the router (and registry) pods use their service accounts for API calls to get client credentials out of envvars

pmorie commented:
+1 @liggitt

On Monday, August 3, 2015, Jordan Liggitt <notifications@github.com> wrote:

> We should also prioritize making the router (and registry) pods use their
> service accounts for API calls to get client credentials out of envvars
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3496#issuecomment-127439522>.
>

smarterclayton commented:
We probably need an endpoint that checks SCC.  Most users won't have access
to SCCs directly and probably shouldn't expect to make calls directly.
Instead, like SAR there should be a PCR (pod creation review)

On Aug 3, 2015, at 8:15 PM, Jordan Liggitt <notifications@github.com> wrote:

The command does validation to make sure the chosen service account has the
required SCC permissions. I'd rather let the pod creation do that
validation, myself (and if errors aren't surfaced clearly enough, work on
improving that). If they want to change the chosen service account after
the fact, it's as simple as editing the router replication controller.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3496#issuecomment-127439081>.

smarterclayton commented:
On that thread, kubectl needs to be able to use pkg/client.InClusterConfig
via some env (either KUBECONFIG=- or similar).  Running kubectl inside a
pod should require exactly one field set by a user, an env.

On Aug 3, 2015, at 8:16 PM, Jordan Liggitt <notifications@github.com> wrote:

We should also prioritize making the router (and registry) pods use their
service accounts for API calls to get client credentials out of envvars

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3496#issuecomment-127439522>.

pweil- commented:
So..is the decision then to remove the check that requires the service account and let the pod fail.  It would fail the SCC check and show up in the events as not having an SCC that allows host ports.  To me, that's a little less obvious than having the command ensure you can use host ports.
